lib/chef/chef_fs/data_handler/data_handler_base.rb

```diff
@@ -64,10 +64,14 @@ def normalize_hash(object, defaults)
           # Make a normalized result in the specified order for diffing
           result = {}
           defaults.each_pair do |key, default|
-            result[key] = object.has_key?(key) ? object[key] : default
+            result[key] = object.is_a?(Hash) && object.key?(key) ? object[key] : default
```

Comment:
This is redundant. The new conditional block does the same thing and add the exception handler. We probably should just remove this block entirely.

Reply:
I am not in agreement with that statement. 🙂 

The normalization is a three step process.
1. Create a new Hash (`results`) to hold the normalized data
2. Iterate over the `defaults` Hash, adding values to `results` but preferring already existing values from `object`
3. Iterate over the `object` Hash, adding values to `results` that don't already exist

We can't change step 2 as it's essential. I've added another commit with a 3rd rspec test showing what the importance of step 2 is: Add items into `result` but prefer the items that already exist in `object` over the default value from `defaults`.

Reply:
@rhass did I make a convincing enough argument to change your mind? :bowtie:

Reply:
Oh, I failed to mention another important part in overviewing! 🤦‍♂️  Previously when the line was just `object.key?(key)` to the left of the `?`  we'd get a `undefined method 'has_key?' for nil:NilClass` error whenever `object` was `nil`

That's because Ruby calls the `has_key?` method on a `nil` object. The NilClass does not have such a method, but it does have the method `.is_a?` (which it inherits from `Object`)

When we combine `object.is_a?(Hash) && object.has_key?(key)` we can guard the right side of the `&&` knowing Ruby evaluates the left side of the `&&` first. When `object` is `nil` `object.is_a?(Hash)` evaluates to `false` so Ruby doesn't try to evaluate `object.has_key?(key)`, assigns the value from the last part of the ternary and doesn't raise a runtime `undefined method` exception.

```
irb(main):005:0> object = nil
=> nil
irb(main):006:0> object.has_key?('foo')
Traceback (most recent call last):
        2: from /opt/chefdk/embedded/bin/irb:11:in `<main>'
        1: from (irb):6
NoMethodError (undefined method `has_key?' for nil:NilClass)
irb(main):007:0> object.is_a?(Hash) && object.has_key?('foo')
=> false
irb(main):008:0>
```